api/coreservice_with_height.go

```diff
@@ -0,0 +1,150 @@
+package api
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/iotexproject/iotex-proto/golang/iotexapi"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/iotexproject/iotex-core/action/protocol"
+	accountutil "github.com/iotexproject/iotex-core/action/protocol/account/util"
+	"github.com/iotexproject/iotex-core/action/protocol/execution/evm"
+	"github.com/iotexproject/iotex-core/blockchain/genesis"
+	"github.com/iotexproject/iotex-core/pkg/tracer"
+)
+
+type (
+	CoreServiceReaderWithHeight interface {
+		Account(addr address.Address) (*iotextypes.AccountMeta, *iotextypes.BlockIdentifier, error)
+	}
+
+	coreServiceReaderWithHeight struct {
+		cs     *coreService
+		height uint64
+	}
+)
+
+func newCoreServiceWithHeight(cs *coreService, height uint64) *coreServiceReaderWithHeight {
+	return &coreServiceReaderWithHeight{
+		cs:     cs,
+		height: height,
+	}
+}
+
+func (core *coreServiceReaderWithHeight) Account(addr address.Address) (*iotextypes.AccountMeta, *iotextypes.BlockIdentifier, error) {
+	ctx, span := tracer.NewSpan(context.Background(), "coreService.Account")
+	defer span.End()
+	addrStr := addr.String()
+	if addrStr == address.RewardingPoolAddr || addrStr == address.StakingBucketPoolAddr {
+		return core.cs.getProtocolAccount(ctx, addrStr)
+	}
+	span.AddEvent("accountutil.AccountStateWithHeight")
+	ctx = genesis.WithGenesisContext(ctx, core.cs.bc.Genesis())
+	stateReader := newStateReaderWithHeight(core.cs.sf, core.height)
+	state, tipHeight, err := accountutil.AccountStateWithHeight(ctx, stateReader, addr)
```

Comment:
the only diff is to use `stateReader` here, which implements archive support
all other codes are duplicate
let's figure out a better way

Reply:
for example, we can define a func to set the height that API wants to access
```
func (cs *coreService) SetHeight(height)
```
and in here
```
if cs.height != 0 {
    sr = newStateReaderWithHeight(core.cs.sf, core.height)
} else {
    sr = core.cs.sf
}
state, tipHeight, err := accountutil.AccountStateWithHeight(ctx, sr, addr)
```

Reply:
discussed offline